zynq: fix, cleanup and update target

luaraneda opened the pull request:
These patches fix a boot problem by reworking the
image creation process, update U-Boot, and switch to
kernel 4.14

Run-tested: ZedBoard

Signed-off-by: Luis Araneda <luaraneda@gmail.com>

luaraneda committed:
zynq: reworked image creation process

After commit a1248da (zynq: convert to new image build code),
all boards, became unbootable, entering into a boot loop.

Replace the compressed kernel zImage by an uncompressed Image,
which is compressed after with gzip. Also, enable the creation
of an initramfs image by default.

Change images' filename to match the compatible string from the
respective .dts file (<vendor>_<model>). Also, use the shared
mkits.sh script and remove the target specific one.

Fixes: FS#1204

Run-tested: ZedBoard

Signed-off-by: Luis Araneda <luaraneda@gmail.com>

luaraneda committed:
zynq: remove hard-coded network configuration

This is no longer necessary since commit 57776e6
(base-files: remove default /etc/config/network,
generate it via board.d instead)

Run-tested: ZedBoard

Signed-off-by: Luis Araneda <luaraneda@gmail.com>

luaraneda committed:
zynq: fix detection of the zybo board

The board_name was being tested for "xlnx,zynq-zybo",
but the .dts compatible string is "digilent,zynq-zybo"

Also, sorted the boards alphabetically, and added an
error message for unsupported boards

Signed-off-by: Luis Araneda <luaraneda@gmail.com>

luaraneda committed:
uboot-zynq: update to 2017.03

Compile-tested: ZedBoard

Signed-off-by: Luis Araneda <luaraneda@gmail.com>

luaraneda committed:
zynq: copy config to kernel 4.14

The file will be used as a base configuration
for kernel 4.14

Signed-off-by: Luis Araneda <luaraneda@gmail.com>

luaraneda committed:
zynq: refresh kernel 4.14 configuration

Signed-off-by: Luis Araneda <luaraneda@gmail.com>

luaraneda committed:
zynq: switch to kernel 4.14

Also, remove the source-only qualifier

Run-tested: ZedBoard

Signed-off-by: Luis Araneda <luaraneda@gmail.com>

luaraneda committed:
zynq: remove support for kernel 4.4

Signed-off-by: Luis Araneda <luaraneda@gmail.com>

commodo commented on target/linux/zynq/config-4.14:
```diff
@@ -0,0 +1,752 @@
+CONFIG_ADVISE_SYSCALLS=y
```
would it be an idea to squash this with `zynq: remove support for kernel 4.4` ? and just do "zynq: move kernel config to 4.14" ?
it would be one less commit; especially if this is a copy

luaraneda replied:
Yes, I did that initially, but decided to split the changes into two commits so the removal could be reverted in case someone needs the 4.4 kernel in the future.

I can squash the changes if you still think is appropriate

commodo replied:
not sure;
let's see if anyone else has any other thoughts; [someone with merge rights]
otherwise you implement the change, and he/she may ask you to leave it like this;

commodo commented:
this looks very nice :)
thanks for the work

txomon commented:
Hey! This works nicely! I have finally been able to run it. however I have noticed a few things.

1. The uboot files have disappeared, making it harder to setup from scratch the board. Before there was the spl folder with all the files to be copied over to the SD. Now I don't have too clear where those files would come from, as compiling in a clean environment is just generating the manifest, the root.tar.gz and the uImage.itb 
2. The default image target is tar.gz instead of the cpio.gz (for the fit image)

Finally, If you don't mind, can you tell which were the kernel config options that were keeping the board from booting? Or was it something that got solved by upgrading the kernel and creating the new config?

luaraneda commented:
Hi,

> 1. The uboot files have disappeared

You have to select the package on the "Boot loaders" section from menuconfig. Once selected, a folder (**u-boot-zed** for the ZedBoard) will have the U-Boot files.
I also did a review of those files, and only two of them are required to boot the board:
+ **boot.bin**: The SPL
+ **u-boot.img**: U-Boot, which is loaded by the SPL

If you need additional files, please let me know so I can add them (they are in the build directory)

> 2. The default image target is tar.gz instead of the cpio.gz (for the fit image)

Do you need a cpio image? I changed that because it was not working on my board (see bellow)

> Finally, If you don't mind, can you tell which were the kernel config options that were keeping the board from booting? Or was it something that got solved by upgrading the kernel and creating the new config?

The kernel configuration was fine. From the commits you can see that just by reworking the image creation process the board started to boot fine. When migrating to kernel 4.14 I changed the kernel configuration, and I had to manually add and remove several configs that are handled automatically by other parts of OpenWRT, like debugfs, initramfs, and nfs. There might still be some of them that should be modified.

I encountered two problems when resolving the booting issue:

+ The kernel image inside the .itb image was a uImage instead of the zImage before the migration. I resolved that easily by removing the **uImage none** step on image/Makefile

+ Once the kernel was initializing, it couldn't mount the rootfs, it panicked with the message

> Initramfs unpacking failed: junk in compressed archive

After trying to resolve that unsuccessfully, I decided to try an image creation process that I was more familiar with, and I borrowed some code from the ipq40xx target, which booted fine.
I think the cpio approach could be fixed. I didn't try long enough, nor I see the advantage of keep trying.

commodo commented:
Hey,
So, one way to address the u-boot package selection is with this change: https://github.com/openwrt/openwrt/pull/695

I also noticed that [since there is no `Default` profile] a multi-profile symbol is added for Zynq [ I don't remember the entire logic.
And for `Default`  profile [`CONFIG_TARGET_<name>_Default` symbol] all u-boot images are selected.
But not for `CONFIG_TARGET_MULTI_PROFILE=y`

txomon commented:
Is there anything left here to do before merging?

luaraneda commented:
I rebased the commits because recent changes have produced a merge conflict:
- f6a067a (kernel: mark all v3.18 and v4.4 kernels are source-only)
- 1289e00 (uboot-zynq: fix build on hosts lacking pkg-config)

@txomon, I don't see anything more to do, from my side at least.

@mkresin Can you merge this? I'm asking because you put the **zynq** label on the PL.
Or do we need more reviews? (I can try to send an e-mail to the **zynq** maintainer).

mybays commented:
spi flash not recognized for u-boot-2017.09

U-Boot 2017.09 (Jun 02 2018 - 01:09:22 +0800)

Model: Zynq Zed Development Board
Board: Xilinx Zynq
DRAM:  ECC disabled 512 MiB
MMC:   sdhci@e0100000: 0
SF: unrecognized JEDEC id bytes: 00, 00, 00
*** Warning - spi_flash_probe_bus_cs() failed, using default environment

In:    serial@e0001000
Out:   serial@e0001000
Err:   serial@e0001000
Model: Zynq Zed Development Board
Board: Xilinx Zynq
Net:   ZYNQ GEM: e000b000, phyaddr 0, interface rgmii-id

Warning: ethernet@e000b000 (eth0) using random MAC address - d6:bb:a6:a9:cb:7a
eth0: ethernet@e000b000
Checking if uenvcmd is set ...
Hit any key to stop autoboot:  0 
Zynq>

luaraneda commented:
> spi flash not recognized for u-boot-2017.09

You are right!
I just checked the last bootlog that I saved and the SPI flash is not recognized.
I must have confused the last fully working U-Boot version, because I tried several versions when debugging the boot problem, and I hit a compilation problem on 2017.11 and newer.
If I remember correctly, the last version that I checked the SPI flash was working is 2017.03

Unfortunately I've returned the ZedBoard that I was using to do the tests.
@mybays, can you test the branch [zynq-update-v3](https://github.com/luaraneda/openwrt/tree/zynq-update-v3) from my repo before I update the PR? I rebased it on master and updated U-Boot to 2017.03

Is there a special reason to use the SPI flash from U-Boot?
I'm aware that the current implementation needs the SPI flash to store the U-Boot environment,
but by doing so you are overriding the SPI flash content, which might booting from the SPI flash.

I'm planning to modify the way U-Boot behaves, and use the file uEnv.txt from the SD-card, and also add support for persistent rootfs, like the sunxi target.
But that will have to wait until I buy a Zybo-Z7, which might be a couple of weeks.

mybays commented:
I have figured out this problem by increasing qspi speed from default 100000 to 3000000.
directly modify drivers/spi/spi-uclass.c,function:int dm_spi_claim_bus(struct udevice *dev)

luaraneda commented:
I haven't look in detail, but that looks like a bug/regression on U-Boot when modifying the driver, or a missing property on the device-tree file.
I would prefer to revert U-Boot to 2017.03 until I have a board to actually test newer versions, unless you need 2017.09
Please test the branch [zynq-update-v3](https://github.com/luaraneda/openwrt/tree/zynq-update-v3) on your board so I can update the PR.

mybays commented:
2017.03 works fine.

luaraneda commented:
PR updated:
+ Rebased against master
+ Changed U-Boot to version to 2017.03 (fix wrong SPI flash initialization)

blogic commented:
merged, Thanks !
